### PR TITLE
DOC: Improved documentation for diagflat, diag and diagonal.

### PR DESCRIPTION
**Pull request by selamw1**
Part of https://github.com/google/jax/issues/21461

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -5667,8 +5696,50 @@ def diagonal(a: ArrayLike, offset: int = 0, axis1: int = 0,
...
+  See also:
+    - :func:`jax.numpy.diagflat`
+    - :func:`jax.numpy.diagonal`
+
+  Examples:
+    Creating a diagonal matrix from a 1-D array:
+
+    >>> jnp.diag(jnp.array([1, 2, 3]))
+    Array([[1, 0, 0],
+            [0, 2, 0],
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5727]: fix spacing

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -5667,8 +5696,50 @@ def diagonal(a: ArrayLike, offset: int = 0, axis1: int = 0,
...
+
+    >>> jnp.diag(jnp.array([1, 2, 3]), k=1)
+    Array([[0, 1, 0, 0],
+            [0, 0, 2, 0],
+            [0, 0, 0, 3],
+            [0, 0, 0, 0]], dtype=int32)
+    
+    Extracting a diagonal from a 2-D array:
+
+    >>> jnp.diag(jnp.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5740]: Same comment as above: consider defining the 2D array separately in a more human-readable way.

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -5685,14 +5756,39 @@ def _diag(v, k):
...
+  """Return a 2-D array with the flattened input array laid out on the diagonal.
+
+  JAX implementation of :func:`numpy.diagflat`.
+
+  This differs from `np.diagflat` for some scalar values of `v`. JAX always returns
+  a two-dimensional array, whereas NumPy may return a scalar depending on the type
+  of `v`.
+  
+  Args:
+    v: Input 1D array.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5769]: I believe this can be N-dimensional, but is always flattened to 1D.

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -5685,14 +5756,39 @@ def _diag(v, k):
...
+    k: optional, int, default=0. Diagonal offset. Positive values place the diagonal
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5770]: Same comment as above regarding parameter descriptions

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -5685,14 +5756,39 @@ def _diag(v, k):
...
+    A 2D array with the input elements placed along the diagonal with the
+    specified offset (k). The remaining entries are filled with zeros.
+
+  See also:
+    - :func:`jax.numpy.diag`
+    - :func:`jax.numpy.diagonal`
+
+  Examples:
+    >>> jnp.diagflat(jnp.array([1, 2, 3]))
+    Array([[1, 0, 0],
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5783]: Fix spacing

File: jax/_src/numpy/lax_numpy.py [hunk 6]
```diff
@@ -5685,14 +5756,39 @@ def _diag(v, k):
...
+            [0, 2, 0],
+            [0, 0, 3]], dtype=int32)
+    >>> jnp.diagflat(jnp.array([1, 2, 3]), k=1)
+    Array([[0, 1, 0, 0],
+            [0, 0, 2, 0],
+            [0, 0, 0, 3],
+            [0, 0, 0, 0]], dtype=int32)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5790]: It could be useful to show what this does for a non-1d input, such as `jnp.array([[1, 2], [3, 4]])`

**Review by jakevdp — COMMENTED** [review]
Looks good! A few comments below

**Comment by selamw1** [comment]
Thanks @jakevdp 
Docstring descriptions, examples and spacing are fixed.

**Review by jakevdp — APPROVED** [review]
Looks good! Triggering the tests.

**Comment by jakevdp** [comment]
There are still some lint errors. You might find it helpful to run the lint check locally, see info here: https://jax.readthedocs.io/en/latest/contributing.html#linting-and-type-checking

**Comment by selamw1** [comment]
> There are still some lint errors. You might find it helpful to run the lint check locally, see info here: https://jax.readthedocs.io/en/latest/contributing.html#linting-and-type-checking

Thanks!  Linting checked locally and fixed.

**Comment by jakevdp** [comment]
Thanks! Can you squash the commits please? Then we'll be able to get this merged

**Pull request merged**